src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -121,99 +346,173 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
-                if (body.Length > 0)
+                    if (_responseType == ResponseType.HeadersOnly)
+                    {
+                        responseServed = true;
+                    }
+                    else if (_responseType == ResponseType.FullReponse)
+                    {
+                        // Copy the cached response body
+                        var body = cachedResponse.Body;
+
+                        // Add a content-length if required
+                        if (response.ContentLength == null && string.IsNullOrEmpty(response.Headers[HeaderNames.TransferEncoding]))
+                        {
+                            response.ContentLength = body.Length;
+                        }
+
+                        if (body.Length > 0)
+                        {
+                            await response.Body.WriteAsync(body, 0, body.Length);
+                        }
+
+                        responseServed = true;
+                    }
+                }
+                else
                 {
-                    await response.Body.WriteAsync(body, 0, body.Length);
+                    // TODO: Validate with endpoint instead
                 }
-                return true;
             }
 
-            return false;
-        }
+            if (!responseServed && RequestCacheControl.OnlyIfCached)
+            {
+                HttpContext.Response.Clear();
+                HttpContext.Response.StatusCode = StatusCodes.Status504GatewayTimeout;
 
-        internal void HookResponseStream()
-        {
-            // TODO: Use a wrapper stream to listen for writes (e.g. the start of the response),
-            // check the headers, and verify if we should cache the response.
-            // Then we should stream data out to the client at the same time as we buffer for the cache.
-            // For now we'll just buffer everything in memory before checking the response headers.
-            // TODO: Consider caching large responses on disk and serving them from there.
-            OriginalResponseStream = HttpContext.Response.Body;
-            Buffer = new MemoryStream();
-            HttpContext.Response.Body = Buffer;
-        }
+                responseServed = true;
+            }
 
-        internal bool OnResponseStarting()
-        {
-            // Evaluate the response headers, see if we should buffer and cache
-            CacheResponse = true; // TODO:
-            return CacheResponse;
+            return responseServed;
         }
 
-        internal void FinalizeCaching()
+        internal void FinalizeCachingHeaders()
         {
             if (CacheResponse)
             {
+                // Create the cache entry now
                 var response = HttpContext.Response;
-                var varyHeaderValue = response.Headers["Vary"];
+                var varyHeaderValue = response.Headers[HeaderNames.Vary];
+                _cachedResponseValidFor = ResponseCacheControl.SharedMaxAge
```

Comment:
This should have a configurable max value. It doesn't need to be huge because the caches have an automatic sliding refresh.